Loading progress bar upgrades

jroweboy opened the pull request:
Part two for #2034, this adds more sparkle to the progress bar to make it more useful. Will rebase this pr after #2034 is merged.

**This image was taken with a loading simulator, and is not representative of actual shader loading times**

![obligatory ui image](https://user-images.githubusercontent.com/952515/51435336-5c0fe980-1c32-11e9-95b9-b540bdc607bd.gif)

There are two problems with the original single progressbar

1. The shader disk loading is actually broken into two parts: loading raw shaders and loading binary shaders.
2. Newly added shaders are appended to the caches, so on the next launch, loading progress goes quickly until it reaches new shaders at which point it'll dramatically slow down.

With these two problems in mind, I designed the progress bar to be minimal but informative. The progressbar is hidden until shader loading is started, and there are two distinct phases with different color progressbars (the colors used in the yuzu icon). If the progress slows down significantly, then we've reached a point where new shaders are either getting decompiled (step 1) or getting compiled (step 2) and we display an estimated time.

Reviewers:
The first 4 commits are from #2034 as this depends on it. Please just review the last commit.

lioncash commented on src/yuzu/loading_screen.cpp:
```diff
@@ -25,11 +28,71 @@
 #endif
 
+/// Static maps defining the differences in text and styling for each stage
+static std::unordered_map<VideoCore::LoadCallbackStage, const char*> stage_translations = {
```
`std::unordered_map`, shouldn't be used in a file-scope context, given it has a non-trivial constructor (and also heap allocates).

lioncash commented on src/yuzu/loading_screen.cpp:
```diff
@@ -25,11 +28,71 @@
+    {VideoCore::LoadCallbackStage::Complete, QT_TRANSLATE_NOOP("LoadingScreen", "Launching...")},
+};
+
+static std::unordered_map<VideoCore::LoadCallbackStage, const char*> progressbar_style = {
```
Ditto

lioncash commented on src/yuzu/loading_screen.h:
```diff
@@ -43,7 +48,11 @@ class LoadingScreen : public QWidget {
     void paintEvent(QPaintEvent* event) override;
 
-    void OnLoadProgress(std::size_t value, std::size_t total);
+public slots:
```
This line isn't needed.

lioncash commented on src/yuzu/loading_screen.h:
```diff
@@ -43,7 +48,11 @@ class LoadingScreen : public QWidget {
 
-    void OnLoadProgress(std::size_t value, std::size_t total);
+public slots:
+    void OnLoadProgress(VideoCore::LoadCallbackStage stage, std::size_t value, std::size_t total);
```
This can technically be private, given it's only used by the LoadingScreen class itself.

jroweboy replied:
Its true, but when we will hook it up with the shader cache, we will need to hook the slot from an external file so I will leave it public.

jroweboy commented:
Review comments addressed. Reminder that #2034 needs to be merged first and I will rebase this one of #2034.

jroweboy commented:
Rebased off master and hid the progress bar when its in Prepare state.

ReinUsesLisp commented on src/yuzu/loading_screen.cpp:
```diff
@@ -26,10 +29,59 @@
+    };
+    progressbar_style = {
+        {VideoCore::LoadCallbackStage::Prepare,
+         R"(
```
IMO this strings would be better as an constexpr outside of the function scope.

jroweboy commented:
* Moved the progress bar style strings into constexpr strings.
* Changed the estimated time so it won't flash if things speed up
* Renamed (Raw, Binary) stages to (Decompile, Build) to try and be more descriptive.

jroweboy commented:
* Changed the background of complete stage to blue. Since the complete stage is meant to be used in "indeterminant" aka idle mode, this should create an effect of red and blue chunks alternating over the progressbar.
* Remove the blue border from around the loading screen
* Set minimum size to 1280, 720 for the widget
* Add a fade out to black effect after the first frame is drawn.

DarkLordZach commented on src/yuzu/loading_screen.cpp:
```diff
@@ -25,11 +30,84 @@
+ background-color: #ff3c28;
+})";
+
+constexpr const char* PROGRESSBAR_STYLE_COMPLETE = R"(
```
For all of these I'm pretty sure you can do `constexpr const char NAME[] = ` instead and it saves space from allocating the pointer.

DarkLordZach commented on src/video_core/rasterizer_interface.h:
```diff
@@ -4,13 +4,22 @@
+enum class LoadCallbackStage {
+    Prepare,
+    Decompile,
+    Build,
```
`BuildShaders` or similar please

DarkLordZach commented on src/video_core/rasterizer_interface.h:
```diff
@@ -4,13 +4,22 @@
 
+enum class LoadCallbackStage {
+    Prepare,
+    Decompile,
```
`DecompileShaders` or similar please
